torch/nn/utils/parametrize.py

```diff
@@ -258,34 +378,86 @@ def right_inverse(self, X: Tensor) -> Tensor
         >>> m.weight = A  # Initialize the weight to be the symmetric matrix A
         >>> print(torch.allclose(m.weight, A))
         True
+
+        >>> class RankOne(nn.Module):
+        >>>     def forward(self, x, y):
+        >>>         # Form a rank 1 matrix multiplying two vectors
+        >>>         return x.unsqueeze(-1) @ y.unsqueeze(-2)
+        >>>
+        >>>     def right_inverse(self, Z):
+        >>>         # Project Z onto the rank 1 matrices
+        >>>         U, S, Vh = torch.linalg.svd(Z, full_matrices=False)
+        >>>         # Return rescaled singular vectors
+        >>>         s0_sqrt = S[0].sqrt().unsqueeze(-1)
+        >>>         return U[..., :, 0] * s0_sqrt, Vh[..., 0, :] * s0_sqrt
+        >>>
+        >>> linear_rank_one = P.register_parametrization(nn.Linear(4, 4), "weight", RankOne())
+        >>> print(torch.linalg.matrix_rank(linear_rank_one.weight).item())
+        1
+
     """
     parametrization.train(module.training)
     if is_parametrized(module, tensor_name):
-        # Just add the new parametrization to the parametrization list
-        module.parametrizations[tensor_name].append(parametrization)  # type: ignore[index, union-attr]
+        # Correctness checks
+        assert isinstance(module.parametrizations, ModuleDict)  # Make mypy happy
+        parametrization_list = module.parametrizations[tensor_name]
+        Y = parametrization_list()
+        Z = parametrization(Y)
+        if Z.dtype != Y.dtype:
+            raise ValueError(
+                "Registering a parametrization may not change the dtype of the parameter.\n"
+                f"self.{tensor_name}.dtype: {Y.dtype}\n"
+                f"parametrization(self.{tensor_name}).dtype: {Z.dtype}"
+            )
+        if Z.shape != Y.shape:
+            raise ValueError(
+                "Registering a parametrization may not change the shape of the parameter.\n"
+                f"self.{tensor_name}.shape: {Y.shape}\n"
+                f"parametrization(self.{tensor_name}).shape: {Z.shape}"
+            )
+        if hasattr(parametrization, "right_inverse"):
+            X = parametrization.right_inverse(Z)  # type: ignore[operator]
+            # nb. The naming of X / Y / Z in the code vs the errors is different
```

Comment:
Why does it have to be?

Reply:
https://github.com/pytorch/pytorch/pull/58488/files/eeb4f204e6fa2e36f3bee06962e3e71053735c4c#r647429832

Reply:
for the one above, I understand that you don't want to use original/new in the sample to keep names short.
But I'm not sure why X and Z are switched here. It would be better to have consistent naming throughout here.